plugins/fontconfig.lua

```diff
@@ -0,0 +1,57 @@
+-- mod-version:1 -- lite-xl 2.00
+local subprocess = require "process"
+
+local core = require "core"
+local style = require "core.style"
+local config = require "core.config"
+
+--[[
+    Example config (put it in user module):
+    ```
+    local fc = require "plugins.fontconfig"
+    fc(
+        { name = "sans", size = 13 * SCALE },     -- UI font
+        { name = "monospace", size = 13 * SCALE } -- code font
+    )
+    ```
+]]
+
+local function resolve_font(spec)
+    local scan_rate = 1 / config.fps
+    local proc = subprocess.start({ "fc-match", "-s", "-f", "%{file}\n", spec }, {
+        stdin = subprocess.REDIRECT_DISCARD,
+        stdout = subprocess.REDIRECT_PIPE,
+        stderr = subprocess.REDIRECT_STDOUT
+    })
+    local prev, lines = {}, {}
+    while proc:running() do
+        coroutine.yield(scan_rate)
+        local buf = proc:read_stdout()
+        local p, _, n = string.match(buf, "(.+)\n(.+)")
+        if p then
+            prev[#prev + 1] = p
+            lines[#lines + 1] = table.concat(prev, "")
+            prev = { n }
+        else
+            prev[#prev + 1] = buf
+        end
+    end
+    table.insert(lines, table.concat(prev, ""))
+
+    if proc:returncode() ~= 0 or #lines < 1 then
+        error(string.format("Cannot find a font matching the given specs: %q", spec), 0)
+    end
+    -- maybe in the future we can detect and do glyph substitution here...
+    return lines[1]
+end
+
+local function load_system_fonts(font, code_font)
+    core.add_thread(function()
```

Comment:
Doing all this asynchronously is very elegant but I am wondering if there is a compelling reason for not doing this synchronously. In my point of view we would have two advantages:

1. code would be simpler
2. it would be more reactive, no need to wait for the thread scheduler.

The only caveat is if the fc-list is extremely long but I guess on a modern system this will never be a problem. I guess reproc reading of stdout plus Lua string matching functions are quite performing even to parse a long list of fonts.

Reply:
It can be much simpler - that I agree. I could have used `io.popen` and read line-by-line. The problem is that this code has to run at startup. On a HDD system, this can be slow.

Reply:
Find point, I didn't consider that it needs to run at startup and the slowlyness we may have due to lot of HD access. I guess the asynchronous implementation you used is the best approach.

Still I am wondering if we may find a solution so that it works for any font, not only UI or code. What if we add a callback function for the async call like below ?

```lua
fontconfig.load_async("sans", 13, function(new_font) style.font = new_font end)
```

This is just an idea to discuss of course. Otherwise instead of the callback function we may provide just a string with the name of the field within "style" that we want to fill with the new font when available. Alternatively it could by something like:

```lua
fontconfig.load {
    font = { name = "sans", size = 13 * SCALE },
    code_font = { name = "monospace", size = 13 * SCALE },
}
```

I like this latter solution because it is readable and uncomplicated.